fem/bilininteg_vecmass.cpp

```diff
@@ -0,0 +1,367 @@
+               for (int qy = 0; qy < Q1D; ++qy)
+               {
+                  const double wy = B(qy,dy);
+                  for (int qx = 0; qx < Q1D; ++qx)
+                  {
+                     sol_xy[qy][qx] += wy * sol_x[qx];
+                  }
+               }
+            }
+            for (int qz = 0; qz < Q1D; ++qz)
+            {
+               const double wz = B(qz,dz);
+               for (int qy = 0; qy < Q1D; ++qy)
+               {
+                  for (int qx = 0; qx < Q1D; ++qx)
+                  {
+                     sol_xyz[qz][qy][qx] += wz * sol_xy[qy][qx];
+                  }
+               }
+            }
+         }
+         for (int qz = 0; qz < Q1D; ++qz)
+         {
+            for (int qy = 0; qy < Q1D; ++qy)
+            {
+               for (int qx = 0; qx < Q1D; ++qx)
+               {
+                  sol_xyz[qz][qy][qx] *= op(qx,qy,qz,e);
+               }
+            }
+         }
+         for (int qz = 0; qz < Q1D; ++qz)
+         {
+            double sol_xy[max_D1D][max_D1D];
+            for (int dy = 0; dy < D1D; ++dy)
+            {
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  sol_xy[dy][dx] = 0;
+               }
+            }
+            for (int qy = 0; qy < Q1D; ++qy)
+            {
+               double sol_x[max_D1D];
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  sol_x[dx] = 0;
+               }
+               for (int qx = 0; qx < Q1D; ++qx)
+               {
+                  const double s = sol_xyz[qz][qy][qx];
+                  for (int dx = 0; dx < D1D; ++dx)
+                  {
+                     sol_x[dx] += Bt(dx,qx) * s;
+                  }
+               }
+               for (int dy = 0; dy < D1D; ++dy)
+               {
+                  const double wy = Bt(dy,qy);
+                  for (int dx = 0; dx < D1D; ++dx)
+                  {
+                     sol_xy[dy][dx] += wy * sol_x[dx];
+                  }
+               }
+            }
+            for (int dz = 0; dz < D1D; ++dz)
+            {
+               const double wz = Bt(dz,qz);
+               for (int dy = 0; dy < D1D; ++dy)
+               {
+                  for (int dx = 0; dx < D1D; ++dx)
+                  {
+                     y(dx,dy,dz,c,e) += wz * sol_xy[dy][dx];
+                  }
+               }
+            }
+         }
+      }
+   });
+}
+
+static void PAVectorMassApply(const int dim,
+                              const int D1D,
+                              const int Q1D,
+                              const int NE,
+                              const Array<double> &B,
+                              const Array<double> &Bt,
+                              const Vector &op,
+                              const Vector &x,
+                              Vector &y)
+{
+   if (dim == 2)
+   {
+      return PAVectorMassApply2D(NE, B, Bt, op, x, y, D1D, Q1D);
```

Comment:
Is this enough? Standard diffusion has optimized cases for the most common orders here. (Just wondering -- I'm not sure what the exact goal of the PR is.)

Reply:
The reason for this is the increased compile time. The users who really need the performance increase will have to define the template parameters (for now).

Reply:
Should we prepare the switch with the templated cases and just #ifdef it out, so the user can just turn the optimized cases on?

Reply:
If we would do that, we have to do it on every other kernel as well to be consistent. It is a good idea and we should consider that.

Reply:
Better yet, the #ifdef could be controlled by the configuration. For development (e.g., "make debug"), you don't want long compile times but for deployment, long compile should be worth the performance improvement...

Reply:
Are you ok to open an issue for that or do you want me to do it? I will bring it up in the next meeting as well.

Reply:
It's just an idea, let's see what the editors think.

Reply:
We decided that we need a separate discussion about how we generally address this. We will keep that going in a new issue, and will merge this PR as is.